src/python/pants/testutil/engine/util.py

```diff
@@ -228,5 +239,139 @@ def fmt_rule(
       f"Get[{product_subject_pair[0]}]({product_subject_pair[1]})"
       for product_subject_pair in gets
     )
-    gets_str = f", gets=[{get_members}]"
+    gets_str = f",\ngets=[{get_members}]"
   return f"@rule({fmt_rust_function(rule)}({params}) -> {product}{gets_str})"
+
+
+def fmt_graph_rule(rule: Callable, *, gets: Optional[List[Tuple[str, str]]] = None) -> str:
+  type_hints = get_type_hints(rule)
+  product = type_hints.pop("return").__name__
+  params = ", ".join(t.__name__ for t in type_hints.values())
+  gets_str = ""
+  if gets:
+    get_members = ',\n'.join(
+      f"Get[{product_subject_pair[0]}]({product_subject_pair[1]})"
+      for product_subject_pair in gets
+    )
+    gets_str = f",\ngets=[{get_members}]"
+  return f"@rule({params}) -> {product}{gets_str}\n{fmt_rust_function(rule)}"
+
+
+@dataclass(frozen=True)
+class RuleFormatRequest:
+  rule: Callable
+  for_param: Optional[Union[Type, Tuple[Type, ...]]] = None
+  gets: Optional[List[Tuple[str, str]]] = None
+
+  def format(self) -> str:
+    msg = fmt_graph_rule(self.rule, gets=self.gets)
+    if self.for_param is not None:
+      if isinstance(self.for_param, type):
+        msg += f'\nfor {self.for_param.__name__}'
+      else:
+        joined = ', '.join(c.__name__ for c in self.for_param)
+        msg += f'\nfor ({joined})'
+    return msg
+
+  @classmethod
+  def format_rule(cls, obj):
+    assert obj is not None
+
+    if isinstance(obj, cls):
+      return obj.format()
+
+    if isinstance(obj, type):
+      return f'Select({obj.__name__})'
+
+    return fmt_graph_rule(obj)
+
+
+def fmt_param_edge(
+    param: Type,
+    product: Union[Type, Tuple[Type, ...]],
+    via_func: Union[Type, RuleFormatRequest],
+    return_func: Optional[RuleFormatRequest] = None,
+) -> str:
+  if isinstance(via_func, type):
+    via_func_str = f'Select({via_func.__name__})'
+  else:
+    via_func_str = RuleFormatRequest.format_rule(via_func)
+
+  if isinstance(product, type):
+    product_name = product.__name__
+  else:
+    joined = ', '.join(p.__name__ for p in product)
+    product_name = f'({joined})'
+
+  return_elements = []
+  if return_func is not None:
+    return_func_str = return_func.format()
+    return_elements.append(return_func_str)
+  return_elements.append(f'Param({param.__name__})')
+  return_str = ' '.join(f'"{el}"' for el in return_elements)
+
+  param_color_fmt_str = RuleType.param.graph_vertex_color_fmt_str()
+  return f"""\
+"Param({param.__name__})" {param_color_fmt_str}    "{via_func_str}
+for {product_name}" -> {{{return_str}}}\
+"""
+
+
+class RuleType(Enum):
```

Comment:
Rather than `RuleType`, would something like `GraphVertex` be more descriptive? I don't think `Rule` describes `param` correctly.